Add emitter and channel support (#87)
eclipse/microprofile-reactive-messaging: spec/src/main/asciidoc/architecture.asciidoc

// ==== Custom message types

// Implementations may provide custom subtypes of `Message` as an end user API, for the purpose of providing additional metadata associated with the messaging provider. If an application declares that it consumes or produces a subtype of the `Message` wrapper, the implementation should validate that the messaging provider supports that message wrapper type, and raise a deployment exception if it doesn't.

> == Use Reactive Messaging in the imperative world

Reviewer: Not sure I agree that this is what we're talking about.

To me, it's more of a "bridge" allowing a developer to receive data from an imperative endpoint, and convert it to a reactive message

Author: Bridging Reactive and Imperative? Feel free to suggest a better heading.

Reviewer: Maybe "Constructing messages from non-reactive code"?

I consider the content of these paragraphs to discuss topics that are irrelevant to what is being done. For instance, the sentence on bridging implies you can mix imperative and reactive programming in a single place, which isn't the case. What is defined is a way to expose messages from imperative code into a reactive stream. It's not a general solution where you can have `@Outgoing` on a JAX-RS Resource method, etc

Author: This feature enables jax-rs resources to publish messages to a topic. JAX-RS is imperative and it is able to connect to reactive world.

Author: I will rephrase some texts. I am not sure about "Constructing messagings from non-reactive code". How about "Sending messages from imperative code to a reactive flow? Any better suggestions welcome?

Author: reworded

> Traditionally, the reactive world and imperative world are separated and operate in parallel. Reactive Messaging deals with streams of data in the reactive world, while the imperative world is pretty much point to point communication. However, imperative programme sometimes needs to connect to reactive streams so that responses can be emitted to a destination service. 

Reviewer 2: point to point and synchronous communication.

Author: done

> It would be nice if Reactive Messaging is available to the imperative world. This section is about making Reactive Messaging available to the imperative world.

Reviewer 2: Not sure about "It would be nice if Reactive Messaging is available to the imperative world."

Reviewer 2: it's so nice that we did it :-D

Author: reworded

The approach is to introduce the concept of _channel_ so that response messages can be published to a channel, which will link to a messaging broker.

=== Connect to a channel

In order to publish messages on a _channel_, we need to connect a channel first. `@Channel("channel_name")` is used for this purpose.
`@Channel` is a qualifier to inject a _channel_ to a CDI bean or any class that supports CDI injection. 

=== Transit messages to a channel

> You can either inject a _Publisher_, _PublisherBuilder_, _Flowable_, with the specified channel, shown below.

Reviewer 2: Flowable is implementation-specific. It should be mentioned in the spec.

Author: removed.

[source, java]
----

@Inject
@Channel("myChannel")
private Publisher<Message<String>> field;

----

The value _myChannel_ indicates the name of the channel. A channel can be used for the following type:

* Publisher<Message<X>> or Publisher<X>
* PublisherBuilder<Message<X>> or PublisherBuilder<X>

> * Flowable<Message<X>> or Flowable<X>

Reviewer 2: Should be removed (see my previous comment)

> In addition to the above messaging publishers, an additional `Emitter` is also providied by this specification. For an instance, in order for a method to send messages, you can use _Emitter_, as demonstrated below.

Reviewer 2: Maybe explain why we have emitters.

If the property does not exist, the default value will be 128 elements. If the buffer is full, an error will be propagated.

The _value_ attribute on `OnOverflow` can be:

* OnOverflow.Strategy.BUFFER - use a buffer, whose size will be determined by the value of _bufferSize_ if set. Otherwise, the size will be the value of the config property _mp.messaging.emitter.defult-buffer-size_ if it exists. Otherwise, 128 will be used. If the buffer is full, an error will be propagated.
* OnOverflow.Strategy.UNBOUNDED_BUFFER - use an unbounded buffer. If the buffer is full, the applicaiton will die of _OutOfMemory_.
* OnOverflow.Strategy.DROP - drops the most recent value if the downstream can’t keep up. It means that new value emitted by the emitter are ignored.
* OnOverflow.Strategy.FAIL - propagates a failure in case the downstream can’t keep up.
* OnOverflow.Strategy.LATEST - keeps only the latest value, dropping any previous value if the downstream can’t keep up.
* OnOverflow.Strategy.NONE - ignore the back-pressure signals letting the downstream consumer to implement a strategy.

Here are some examples:

[source, java]
----

@Inject
@Channel("myChannel")

> @OnOverflow(value = OnOverflow.Strategy.BUFFER) // Buffer strategy using the buffer size specified by _mp.messaging.emitter.defult-buffer-size_ if exists. Otherwise, 128 will be used.

Reviewer 2: default (a is missing)

Reviewer 2: @Emily-Jiang This one has not been fixed.

Author: fixed

Emitter<String> emitter;

@Inject
@Channel("myChannel") // Buffer strategy will be used. It behaves as if @OnOverflow(value = OnOverflow.Strategy.BUFFER) is present.
Emitter<String> emitter;

@Inject
@Channel("myChannel")
@OnOverflow(value = OnOverflow.Strategy.DROP) // Drop the most recent values
Emitter<String> emitter;
----

>     

Reviewer 2: too many blank spaces.

Author: deleted

== Publishing messages to a stream from imperative code
Traditionally, the reactive world and imperative world are separated and operate in parallel. Reactive Messaging deals with streams of data in the reactive world, while the imperative world is pretty much point to point and synchronous communication. However, imperative programme sometimes needs to connect to reactive streams so that responses can be emitted to a destination service. 
Bridging the two worlds is very valuable thing to do, so that one microservice can use technologies from both environments. For an instance, a JAX-RS resource might want to publish messages to a Reactive Messaging channel. This section is about enabling imperative code to publish messages to a Reactive Messaging channel, so that it can be consumed by a consumer.

> The approach is to introduce the concept of _channel_.

Reviewer 2: Channel is a concept in the spec. It's not something introduced here. What is introduced in `@Channel`.

Author: done

> === Connect to a channel

Reviewer 2: Retrieving a channel

Author: I prefer connect over retrieve, as retrieving means consuming things instead of publishing stuff. How about change to _Connecting to a channel_

Author: or Eastablish a channel?

In order to publish messages on a _channel_, we need to connect a channel first. `@Channel("channel_name")` is used for this purpose.

> `@Channel` is a qualifier to inject a _channel_ to a CDI bean or any class that supports CDI injection. 

Reviewer 3: I don't think this paragraph is necessary. From reading the later paragraphs, it seems like you can inject `Publisher` or `Emitter` and you use the `@Channel` annotation to say which channel it should be connected to.

I think it's easier to understand when it's presented in the context of injecting one of the other objects, rather than declaring it up front.

Author: ok. Moving that paragraph to the following section.

> === Transit messages to a channel

Reviewer 3: You can't use "Transit" like this. "Transit messages through a channel" might make sense, but it would be better not to use the word Transit here.

I assume we're talking about sending and receiving messages via channels. It might make more sense to break up this section with two headings.

=== Receiving messages from a channel.

=== Sending messages to a channel

Author: changed to sending messaging to a channel

> You can either inject a _Publisher_ or _PublisherBuilder_ with the specified channel, shown below.

Reviewer 3: ```suggestion
You can inject either a `Publisher` or `PublisherBuilder` for the specified channel, as shown below.
```

[source, java]
----

@Inject
@Channel("myChannel")
private Publisher<Message<String>> field;

----

The value _myChannel_ indicates the name of the channel. A channel can be used for the following type:

> * Publisher<Message<X>> or Publisher<X>

Reviewer 2: Ticks missing

Author: done

> * PublisherBuilder<Message<X>> or PublisherBuilder<X>

Reviewer 2: Ticks missing

Author: done

> * `PublisherBuilder<Message<X>>` or `PublisherBuilder<X>`

Reviewer 3: This section talks about injecting a `Publisher` or `PublisherBuilder` but doesn't say why you're doing this.

I assume it's so that the user can subscribe and receive messages? It might be good to spell that out. An additional example showing how to use the injected object might be helpful here too.

I'm also not sure that this actually helps bridge between reactive and non-reactive code since you still just have a reactive object. In that case, it might be better if it was moved out of the section `Publishing messages to a stream from imperative code`.

Author: reworked

> In addition to the above messaging publishers, an additional `Emitter` is also providied by this specification. 

Reviewer 2: Typo in providied.

Author: fixed.

> * Emitter<X>

Reviewer 2: Ticks missing

Author: done

> In addition to the above messaging publishers, an additional `Emitter` is also provided by this specification. 
> 
> * `Emitter<X>`

Reviewer 3: ```suggestion
In addition to the above messaging publishers, the `Emitter` interface is also provided by this specification. 

* `Emitter<X>`
```

Author: reworked

> For an instance, in order for a method to send messages, you can use _Emitter_ to send payload `X` or `Message<X>`, as demonstrated below.

Reviewer 3: ```suggestion
You can inject an `Emitter` and use it to send both payloads (`X`) and messages (`Message<X>`) to a channel as demonstrated below.
```

Author: done

[source, java]
----
@Inject @Channel("myChannel")
private Emitter<String> emitter;

public void publishMessage() {
  emitter.send("a");
  emitter.send("b");
  emitter.complete();
}
----

[source, java]
----
@Inject
@Channel("foo")
private Emitter<String> emitter;

public void run() {
    emitter.send(Message.of("a"));
    emitter.send(Message.of("b"));
    emitter.send(Message.of("c"));
}

----

When injecting an _Emitter_ (e.g. `@Inject @Emitter`), you can configure how the backpressure is handled. The `@OnOverflow` annotation configures how the backpressure behaves when emitting messages/payloads faster than the consumption of the messages.

[source, java]
----
@Inject @Channel("myChannel")
@OnOverflow(value=OnOverflow.Strategy.BUFFER, bufferSize=300)

> Private Emitter<String> emitter;

Reviewer 3: ```suggestion
private Emitter<String> emitter;
```

public void publishMessage() {
  emitter.send("a");
  emitter.send("b");
  emitter.complete();
}
----

> In the above snippet, the buffer size is set to 300 elements. If `_@OnOverflow_` is absent, the buffer strategy _OnOverflow.Strategy.BUFFER_ will be used. 

Reviewer 3: ```suggestion
In the above snippet, the buffer size is set to 300 elements. If `@OnOverflow` is absent, the buffer strategy `OnOverflow.Strategy.BUFFER` will be used.
```

Author: fixed

> If the `bufferSize` is not specified, the size will be the value of the config property _mp.messaging.emitter.default-buffer-size_.

Reviewer 3: ```suggestion
If the `bufferSize` is not specified, the size will be the value of the config property `mp.messaging.emitter.default-buffer-size`.
```

Author: updated

If the property does not exist, the default value will be 128 elements. If the buffer is full, an error will be propagated.

The _value_ attribute on `OnOverflow` can be:

> * `OnOverflow.Strategy.BUFFER` - use a buffer, whose size will be determined by the value of _bufferSize_ if set. Otherwise, the size will be the value of the config property _mp.messaging.emitter.defult-buffer-size_ if it exists. Otherwise, 128 will be used. If the buffer is full, an error will be propagated.

Reviewer 2: typo in `defult`

Author: fixed

> * `OnOverflow.Strategy.UNBOUNDED_BUFFER` - use an unbounded buffer. If the buffer is full, the applicaiton will die of _OutOfMemory_.

Reviewer 2: typo in application

Author: done

> * `OnOverflow.Strategy.UNBOUNDED_BUFFER` - use an unbounded buffer. If the buffer is full, the application will die of _OutOfMemory_.

Reviewer 3: ```suggestion
* `OnOverflow.Strategy.UNBOUNDED_BUFFER` - use an unbounded buffer. The application may run out of memory if values are continually added faster than they are consumed.
```
Matching suggested javadoc change.

Author: fixed.

* `OnOverflow.Strategy.DROP` - drops the most recent value if the downstream can’t keep up. It means that new value emitted by the emitter are ignored.
* `OnOverflow.Strategy.FAIL` - propagates a failure in case the downstream can’t keep up.
* `OnOverflow.Strategy.LATEST` - keeps only the latest value, dropping any previous value if the downstream can’t keep up.

> * `OnOverflow.Strategy.NONE` - ignore the back-pressure signals letting the downstream consumer to implement a strategy.

Reviewer 2: ignores

Author: done

Here are some examples:

[source, java]
----

@Inject
@Channel("myChannel")
@OnOverflow(value = OnOverflow.Strategy.BUFFER) // Buffer strategy using the buffer size specified by _mp.messaging.emitter.default-buffer-size_ if exists. Otherwise, 128 will be used.
Emitter<String> emitter;

@Inject
@Channel("myChannel") // Buffer strategy will be used. It behaves as if @OnOverflow(value = OnOverflow.Strategy.BUFFER) is present.
Emitter<String> emitter;

@Inject
@Channel("myChannel")
@OnOverflow(value = OnOverflow.Strategy.DROP) // Drop the most recent values
Emitter<String> emitter;
----
    
Since the `@Channel("myChannel")` is used to produce messages, a consumer with the `@Incoming("myChannel")` should be specified to consume the messages transiting on the channel _myChannel_.

> The `IllegalStateException` will be thrown if no consumer is found by the time a message is emitted to a channel.

Reviewer 3: ```suggestion
An `IllegalStateException` will be thrown if no consumer is found by the time a message is emitted to the channel.
```

Bridging the two worlds is very valuable thing to do, so that one microservice can use technologies from both environments. For an instance, a JAX-RS resource might want to publish messages to a Reactive Messaging channel. This section is about enabling imperative code to publish messages to a Reactive Messaging channel, so that it can be consumed by a consumer.

You can inject an `Emitter` and use it to send either payloads (`X`) or messages (`Message<X>`) to a channel as demonstrated below.

[source, java]
----
@Inject @Channel("myChannel")
private Emitter<String> emitter;

public void publishMessage() {
  emitter.send("a");
  emitter.send("b");
  emitter.complete();
}
----

[source, java]
----
@Inject
@Channel("foo")
private Emitter<String> emitter;

public void run() {
    emitter.send(Message.of("a"));
    emitter.send(Message.of("b"));
    emitter.send(Message.of("c"));
}

----

> When injecting an `Emitter` (e.g. `@Inject Emitter`), you can configure how the back pressure is handled via `@OnOverflow` annotation, for the situation where emitting messages/payloads faster than the consumption of the messages.

Reviewer 3: This whole section is now way better :+1:.

I'd just recommend this one addition to actually state that the `@Channel` qualifier must be used. It's implied by all the examples but unless I've missed it somewhere, we don't actually state it explicitly.

```suggestion
When injecting an `Emitter`, you must specify the target channel name using the `@Channel` qualifier.

When injecting an `Emitter`, you can configure how the back pressure is handled via `@OnOverflow` annotation, for the situation where emitting messages/payloads faster than the consumption of the messages.
```

Author: rephrased.

== Publishing messages to a channel from imperative code

Traditionally, the reactive world and imperative world are separated and operate in parallel. Reactive Messaging deals with streams of data in the reactive world, while the imperative world is pretty much point to point and synchronous communication. However, imperative programme sometimes needs to connect to reactive streams so that responses can be emitted to a destination service. 
Bridging the two worlds is very valuable thing to do, so that one microservice can use technologies from both environments. For an instance, a JAX-RS resource might want to publish messages to a Reactive Messaging channel. This section is about enabling imperative code to publish messages to a Reactive Messaging channel, so that it can be consumed by a consumer.

This is achieved by using `@Channel`.

=== Sending messages to a channel

In order to publish messages on a _channel_, we need to connect a channel first. `@Channel("channel_name")` is used for this purpose.

> `@Channel` is a qualifier to inject a _channel_ to a CDI bean or any class that supports CDI injection. 

Reviewer 3: "Connect a channel" doesn't read right. You could connect to a channel, or connect something else to a channel.

Also, a "channel" isn't something you inject. You either inject a `Publisher`, `PublisherBuilder` or `Emitter`.

So maybe remove these two lines and incorporate this information into the sections below:

 You can either inject a `Publisher` or `PublisherBuilder` to receive messages from a specified channel, as shown below. You must use the `@Channel` qualifier to specify the name of the channel.

You can inject an `Emitter` and use it to send either payloads (`X`) or messages (`Message<X>`) to a channel as demonstrated below. When injecting the `Emitter` you must use the `@Channel` qualifier to specify the name of the channel.

Author: Thanks @Azquelt ! reworded with your help.

You can either inject a `Publisher` or `PublisherBuilder` with the specified channel, as shown below.

> 

Reviewer 3: The heading is "Sending messages to a channel" but then goes on the show how to _receive_ messages from a channel by injecting a `Publisher`.

Author: fixed
